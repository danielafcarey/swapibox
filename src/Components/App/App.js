import React, { Component } from 'react';
import './App.css';

import Navigation from '../Navigation/Navigation.js';
import Landing from '../Landing/Landing.js';
import People from '../People/People.js';
import Planets from '../Planets/Planets.js';
import Vehicles from '../Vehicles/Vehicles.js';
import Favorites from '../Favorites/Favorites.js';

import ApiHelper from '../../Api.js';
const apiHelper = new ApiHelper();

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedButton: '',
      selectedData: [],
      crawlData: {},
      favorites:  []
    }
  }

  // componentDidMount = async () => {
  //   const crawlData = await apiHelper.getCrawlData();

  //   this.setState({ crawlData })
  // }

  changeCategory = async (event) => {
    const buttonName = event.target.name;

    if (buttonName === 'People') {
      await this.getPeopleData();
    } else if (buttonName === 'Planets') {
      await this.getPlanetsData();
    } else if (buttonName === 'Vehicles') {
      await this.getVehiclesData();
    } else if (buttonName === 'Favorites') {
      await this.getCardsDisplay();
    }

    await this.changeSelectedButton(buttonName);
  }

  changeSelectedButton = (selectedButton) => {
    this.setState({ selectedButton });
  }

  changeDataState = (selectedData) => {
    this.setState({ selectedData });
  }

  getPeopleData = async () => {
    const peopleData = await apiHelper.getPeopleData();
    this.compareToFavorites(peopleData);
  }

  getPlanetsData = async () => {
    const planetsData = await apiHelper.getPlanetsData();
    this.compareToFavorites(planetsData);
  }

  getVehiclesData = async () => {
    const vehiclesData = await apiHelper.getVehiclesData();
    this.compareToFavorites(vehiclesData);
  }

  toggleFavorite = (cardId) => {
    const newSelectedData = this.state.selectedData.map(card => {
      if (cardId === card.id) {
        card.favorite = !card.favorite
      }

      return card;
    });

    this.setState(
      { selectedData: newSelectedData }, 
      () => this.updateFavoritesInState(cardId) 
    )
  }

  updateFavoritesInState = (cardId) => {
    const selectedCard = this.state.selectedData.find(data => {
      console.log(data, data.id, cardId);
      return data.id === cardId;
    });

    if (!selectedCard) {
      this.removeFromFavorites(cardId);
      return;
    }

    if (selectedCard.favorite) {
      this.setState({ favorites: [...this.state.favorites, selectedCard ] });
    } else {
      this.removeFromFavorites(cardId);
    }
  }

  compareToFavorites = (selectedData) => {
    const newSelectedData = selectedData.map(data => {
      if (this.state.favorites.find(favorite => favorite.id === data.id)) {
        data.favorite = true;
      }
      
      return data;
    }) 

    this.changeDataState(newSelectedData);
  }

  removeFromFavorites = (cardId) => {
    const newFavorites = this.state.favorites.filter(favorite => {
      return favorite.id !== cardId
    })

    this.setState({ favorites: newFavorites })
  }

  getDisplayElements = () => {
    if (Object.keys(this.state.selectedData).length === 0) {
      return <Landing crawlData={ this.state.crawlData }/> 
    } else {
      return this.getCardsDisplay();        
    }
  }

  getCardsDisplay = () => {
    const { selectedButton, selectedData, favorites } = this.state;
    if (selectedButton === 'People' && selectedData[0].homeworld) {
      return (
        <People 
          cardData={ selectedData } 
          toggleFavorite={ this.toggleFavorite }  
        />
      );    
    } else if (selectedButton === 'Planets' && selectedData[0].climate) {
      return (
        <Planets 
          cardData={ selectedData }
          toggleFavorite={ this.toggleFavorite } 
        />
      );
    } else if (selectedButton === 'Vehicles' && selectedData[0].model) {
      return (
        <Vehicles 
          cardData={ selectedData } 
          toggleFavorite={ this.toggleFavorite } 
        />
      );
    } else if (selectedButton === 'Favorites') {
      return (
        <Favorites 
          cardData={ favorites } 
          toggleFavorite={ this.toggleFavorite }
        />
      );
    }
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">swapi box</h1>
          <Navigation 
            changeCategory={ this.changeCategory }
            selectedButton={ this.state.selectedButton }/>
        </header>
        { this.getDisplayElements() } 
      </div>
    );
  }
}

export default App;
