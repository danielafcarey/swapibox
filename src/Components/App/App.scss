@import '../Navigation/Navigation.scss';
@import '../Landing/Landing.scss';
@import '../People/People.scss';
@import '../Planets/Planets.scss';
@import '../Vehicles/Vehicles.scss';
@import '../Favorites/Favorites.scss';
@import '../Card/Card.scss';

@font-face {
  font-family: starWarsOutlined;
  src: url(../../Starjhol.ttf);
}

header {
  background-color: black;
  z-index: 1;
}

.App-title {
  display: inline-block;
  color: white;
  font-family: starWarsOutlined;
  font-size: 4rem;
  margin: 0 20px;
  line-height: 5rem;
}

.card-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  width: 85%;
  margin: 75px auto;
}

