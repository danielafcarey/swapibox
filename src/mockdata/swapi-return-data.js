const peopleReturn = [{"homeworld":"Tatooine","homeworldPop":"200000","name":"Luke Skywalker","species":["Human"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"C-3PO","species":["Droid"],"favorite":false},{"homeworld":"Naboo","homeworldPop":"4500000000","name":"R2-D2","species":["Droid"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"Darth Vader","species":["Human"],"favorite":false},{"homeworld":"Alderaan","homeworldPop":"2000000000","name":"Leia Organa","species":["Human"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"Owen Lars","species":["Human"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"Beru Whitesun lars","species":["Human"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"R5-D4","species":["Droid"],"favorite":false},{"homeworld":"Tatooine","homeworldPop":"200000","name":"Biggs Darklighter","species":["Human"],"favorite":false},{"homeworld":"Stewjon","homeworldPop":"unknown","name":"Obi-Wan Kenobi","species":["Human"],"favorite":false}];

const planetsReturn = [{"name":"Alderaan","terrain":"grasslands, mountains","population":"2000000000","climate":"temperate","residents":["Leia Organa","Bail Prestor Organa","Raymus Antilles"],"favorite":false},{"name":"Yavin IV","terrain":"jungle, rainforests","population":"1000","climate":"temperate, tropical","residents":[],"favorite":false},{"name":"Hoth","terrain":"tundra, ice caves, mountain ranges","population":"unknown","climate":"frozen","residents":[],"favorite":false},{"name":"Dagobah","terrain":"swamp, jungles","population":"unknown","climate":"murky","residents":[],"favorite":false},{"name":"Bespin","terrain":"gas giant","population":"6000000","climate":"temperate","residents":["Lobot"],"favorite":false},{"name":"Endor","terrain":"forests, mountains, lakes","population":"30000000","climate":"temperate","residents":["Wicket Systri Warrick"],"favorite":false},{"name":"Naboo","terrain":"grassy hills, swamps, forests, mountains","population":"4500000000","climate":"temperate","residents":["R2-D2","Palpatine","Jar Jar Binks","Roos Tarpals","Rugor Nass","Ric Olié","Quarsh Panaka","Gregar Typho","Cordé","Dormé","Padmé Amidala"],"favorite":false},{"name":"Coruscant","terrain":"cityscape, mountains","population":"1000000000000","climate":"temperate","residents":["Finis Valorum","Adi Gallia","Jocasta Nu"],"favorite":false},{"name":"Kamino","terrain":"ocean","population":"1000000000","climate":"temperate","residents":["Boba Fett","Lama Su","Taun We"],"favorite":false},{"name":"Geonosis","terrain":"rock, desert, mountain, barren","population":"100000000000","climate":"temperate, arid","residents":["Poggle the Lesser"],"favorite":false}];

const vehiclesReturn = [{"name":"Sand Crawler","model":"Digger Crawler","vehicleClass":"wheeled","numberOfPassengers":"30","favorite":false},{"name":"T-16 skyhopper","model":"T-16 skyhopper","vehicleClass":"repulsorcraft","numberOfPassengers":"1","favorite":false},{"name":"X-34 landspeeder","model":"X-34 landspeeder","vehicleClass":"repulsorcraft","numberOfPassengers":"1","favorite":false},{"name":"TIE/LN starfighter","model":"Twin Ion Engine/Ln Starfighter","vehicleClass":"starfighter","numberOfPassengers":"0","favorite":false},{"name":"Snowspeeder","model":"t-47 airspeeder","vehicleClass":"airspeeder","numberOfPassengers":"0","favorite":false},{"name":"TIE bomber","model":"TIE/sa bomber","vehicleClass":"space/planetary bomber","numberOfPassengers":"0","favorite":false},{"name":"AT-AT","model":"All Terrain Armored Transport","vehicleClass":"assault walker","numberOfPassengers":"40","favorite":false},{"name":"AT-ST","model":"All Terrain Scout Transport","vehicleClass":"walker","numberOfPassengers":"0","favorite":false},{"name":"Storm IV Twin-Pod cloud car","model":"Storm IV Twin-Pod","vehicleClass":"repulsorcraft","numberOfPassengers":"0","favorite":false},{"name":"Sail barge","model":"Modified Luxury Sail Barge","vehicleClass":"sail barge","numberOfPassengers":"500","favorite":false}];


export {
  peopleReturn,
  planetsReturn,
  vehiclesReturn
}
