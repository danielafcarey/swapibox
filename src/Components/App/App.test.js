import React from 'react';
import { shallow, mount } from 'enzyme';
import App from './App';
import ApiHelper from '../../Api.js';
const apiHelper = new ApiHelper();
import { peopleReturn, planetsReturn, vehiclesReturn } from '../../mockdata/swapi-return-data.js';

jest.mock('../../Api.js', () => {
  return jest.fn().mockImplementation(() => {
    return {
      getCrawlData: jest.fn().mockImplementation(() => {
        return {
          crawlText: '',
          title: '',
          releaseDate: ''
        }
      }),
      getPeopleData: jest.fn().mockImplementation(() => Promise.resolve(['i am a person'])), 
      getVehiclesData: jest.fn().mockImplementation(() => Promise.resolve(['i am a vehicle'])), 
      getPlanetsData: jest.fn().mockImplementation(() => Promise.resolve(['i am a planet'])) 
    };
  });
});

describe('App', () => {
  let app;
  let appInst;

  beforeEach(() => {
    app = shallow(<App />)
    appInst = app.instance();
  })

  it('matches the snapshot', () => {
    expect(app).toMatchSnapshot();
  });

  it('sets the crawl data in state', async () => {

    expect(app.state('crawlData')).toHaveProperty('crawlText');
    expect(app.state('crawlData')).toHaveProperty('title');
    expect(app.state('crawlData')).toHaveProperty('releaseDate');
  })

  it('updates the loading state when a button is clicked', () => {
    const mockEvent = { target: { name: 'People' } };
    appInst.getPeopleData = jest.fn();

    expect(app.state('loading')).toEqual(false);

    appInst.changeCategory(mockEvent);

    expect(app.state('loading')).toEqual(true);
  })

  it('updates the loading state back to false if favorites button clicked', () => {
    const mockEvent = { target: { name: 'Favorites' } };
    appInst.getCardsDisplay = jest.fn();

    appInst.changeCategory(mockEvent);

    expect(app.state('loading')).toEqual(false);
  })
 
  it('calls getPeopleData if the people button is clicked', () => {
    const mockEvent = { target: { name: 'People' } };
    appInst.getPeopleData = jest.fn();

    appInst.changeCategory(mockEvent);
    
    expect(appInst.getPeopleData).toHaveBeenCalled();
  })

  it('calls getPlanetsData if the planets button is clicked', () => {
    const mockEvent = { target: { name: 'Planets' } };
    appInst.getPlanetsData = jest.fn();

    appInst.changeCategory(mockEvent);

    expect(appInst.getPlanetsData).toHaveBeenCalled();
  })

  it('calls getVehiclesData if the vehicles button is clicked', () => {
    const mockEvent = { target: { name: 'Vehicles' } };
    appInst.getVehiclesData = jest.fn();

    appInst.changeCategory(mockEvent);

    expect(appInst.getVehiclesData).toHaveBeenCalled();
  })

  it('calls getCardsDisplay if the favorites button is clicked', () => {
    const mockEvent = { target: { name: 'Favorites' } };
    appInst.getCardsDisplay = jest.fn();

    appInst.changeCategory(mockEvent);

    expect(appInst.getCardsDisplay).toHaveBeenCalled();
  })

  it('updates selectedButton in App state when button is selected', () => {
    const expected = 'People';

    appInst.changeSelectedButton('People');

    expect(app.state('selectedButton')).toEqual('People');
  })

  it('updates selectedData in state', () => {
    const mockSelectedData = [];
    
    appInst.changeDataState(mockSelectedData);

    expect(app.state('selectedData')).toEqual(mockSelectedData); 
  })

  it('calls compareToFavorites with peopleData', async () => {
    appInst.compareToFavorites = jest.fn();
    const mockPeopleData = ['i am a person']; 

    await appInst.getPeopleData();

    expect(appInst.compareToFavorites).toHaveBeenCalledWith(mockPeopleData);
  })

  it('calls compareToFavorites with planetsData', async () => {
    appInst.compareToFavorites = jest.fn();
    const mockPlanetsData = ['i am a planet']; 

    await appInst.getPlanetsData();

    expect(appInst.compareToFavorites).toHaveBeenCalledWith(mockPlanetsData);
  })

  it('calls compareToFavorites with vehiclesData', async () => {
    appInst.compareToFavorites = jest.fn();
    const mockVehiclesData = ['i am a vehicle']; 

    await appInst.getVehiclesData();

    expect(appInst.compareToFavorites).toHaveBeenCalledWith(mockVehiclesData);
  })

  it('toggles the favorite key on a card in state', () => {
    const prevState = [{ id: 1, favorite: false }]
    const expectedState = [{ id: 1, favorite: true }]
    app.setState({ selectedData: prevState })

    expect(app.state('selectedData')).toEqual(prevState);

    appInst.toggleFavorite(1);

    expect(app.state('selectedData')).toEqual(expectedState);
  })

  it('calls updateFavoritesInState with the cardId after toggling favorite in state', () => {
    appInst.updateFavoritesInState = jest.fn();
    const prevState = [{ id: 1, favorite: false }];
    const expectedState = [{ id: 1, favorite: true }];
    app.setState({ selectedData: prevState });

    appInst.toggleFavorite(1);

    expect(app.state('selectedData')).toEqual(expectedState);
    expect(appInst.updateFavoritesInState).toHaveBeenCalledWith(1);
  })

  it('calls removeFromFavorites with cardId if card was not found in selectedData', () => {
    const mockFavoriteCard = { id: 1, favorite: true };
    appInst.setState({ favorites: [mockFavoriteCard] });
    appInst.removeFromFavorites = jest.fn();

    appInst.updateFavoritesInState(1);

    expect(appInst.removeFromFavorites).toHaveBeenCalledWith(1);
  })

  it('adds a card to favorites if it is marked as a favorite in selectedData', () => {
    const mockSelectedData = [{ id: 2, favorite: true }];
    appInst.setState({ favorites: [], selectedData: mockSelectedData });

    appInst.updateFavoritesInState(2);

    expect(app.state('favorites')).toEqual(mockSelectedData);
  })

  it('calls removeFromFavorites with the cardId if it is not marked as a favorite in selectedData', () => {
    const mockSelectedData = [{ id: 2, favorite: false }];
    appInst.setState({ selectedData: mockSelectedData });
    appInst.removeFromFavorites = jest.fn();

    appInst.updateFavoritesInState(2);

    expect(appInst.removeFromFavorites).toHaveBeenCalledWith(2);
  })

  it('changes the value of favorite for a new data object and passes that to changeDataState', () => {
    const mockFavoriteCard = { id: 1, favorite: true };
    appInst.setState({ favorites: [mockFavoriteCard] });
    appInst.changeDataState = jest.fn();
    const mockSelectedData = [{ id: 1, favorite: false }];
    const expected = [mockFavoriteCard];

    appInst.compareToFavorites(mockSelectedData);

    expect(appInst.changeDataState).toHaveBeenCalledWith(expected);
  })

  it('removes a card from the favorites in state', () => {
    const mockFavoritesBefore = [{ id: 1, favorite: true }, {id: 2, favorite: true }];
    const expectedFavoritesAfter = [{ id: 1, favorite: true }];
    appInst.setState({ favorites: mockFavoritesBefore });

    appInst.removeFromFavorites(2);

    expect(app.state('favorites')).toEqual(expectedFavoritesAfter);
  })

  it('calls getCardsDisplay if there is selectedData in state', () => {
    appInst.getCardsDisplay = jest.fn();
    app.setState({ selectedData: [{ id: 1 }] });

    appInst.getDisplayElements();

    expect(appInst.getCardsDisplay).toHaveBeenCalled();
  })

  it('matches the snapshot if People has been selected', () => {
    app.setState({ selectedData: peopleReturn, selectedButton: 'People' });

    expect(app).toMatchSnapshot();
  })
  it('matches the snapshot if Planets has been selected', () => {
    app.setState({ selectedData: planetsReturn, selectedButton: 'Planets' });

    expect(app).toMatchSnapshot();
  })

  it('matches the snapshot if Vehicles has been selected', () => {
    app.setState({ selectedData: vehiclesReturn, selectedButton: 'Vehicles' });

    expect(app).toMatchSnapshot();
  })

  it('matches the snapshot if Favorites has been selected', () => {
    app.setState({ favorites: peopleReturn, selectedButton: 'Favorites' });

    expect(app).toMatchSnapshot();
  })



})

