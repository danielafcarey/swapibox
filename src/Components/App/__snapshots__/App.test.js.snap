// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="App-header"
>
          <h1
                    className="App-title"
          >
                    swapi box
          </h1>
          <Navigation
                    changeCategory={[Function]}
          />
</header>,
        <Landing />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="App-title"
>
              swapi box
</h1>,
            <Navigation
              changeCategory={[Function]}
/>,
          ],
          "className": "App-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "swapi box",
              "className": "App-title",
            },
            "ref": null,
            "rendered": "swapi box",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "changeCategory": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="App-header"
>
            <h1
                        className="App-title"
            >
                        swapi box
            </h1>
            <Navigation
                        changeCategory={[Function]}
            />
</header>,
          <Landing />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="App-title"
>
                swapi box
</h1>,
              <Navigation
                changeCategory={[Function]}
/>,
            ],
            "className": "App-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "swapi box",
                "className": "App-title",
              },
              "ref": null,
              "rendered": "swapi box",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "changeCategory": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
